dapp/contracts/Crowdsale.sol

```diff
@@ -0,0 +1,210 @@
+        Finished
+    }
+
+    State public state;
+    GetToken public token;
+
+    uint256 public tokensLeft;
+    uint256 public tokensGoal;
+    uint256 public startDate;
+    uint256 public endDate;
+
+    event StateTransition(uint256 _oldState, uint256 _newState);
+    event StateTimeout(uint256 _oldState);
+
+    function nextState(State _newState) private {
+        StateTransition(uint256(state), uint256(_newState));
+        state = _newState;
+    }
+
+    function processTimeouts() private {
+        if (state == State.Started) {
+            if (block.timestamp > endDate) {
+                StateTimeout(uint256(state));
+                nextState(State.Finished);
+            }
+        }
+    }
+
+    function Crowdsale(GetToken _token, IBosonAddressResolver _bar) public {
+        token = _token;
+        bosonResolver = _bar;
+        owner = msg.sender;
+        state = State.Initialized;
+    }
+
+    function start() onlyOwner public {
+        require(state == State.Initialized);
+
+        state = State.Started;
+        startDate = block.timestamp;
+        endDate = startDate + DURATION;
+        tokensGoal = token.balanceOf(this);
+        tokensLeft = tokensGoal;
+
+        require(tokensGoal > 0);
+        require(price() > 0);
+    }
+
+    function price() public view returns (uint256 _price) {
+        require(state >= State.Started);
+        return GOAL_WEI / tokensGoal;
+    }
+
+    function buy(address buyer) public payable {
+        processTimeouts();
+        require(state == State.Started);
+        require(buyer != 0x0);
+        require(approved(buyer));
+
+        // Convert purchase to requested tokens.
+        uint256 tokens = msg.value.div(price());
+
+        // Make sure to not sell more than we have.
+        if (tokensLeft < tokens) {
+            tokens = tokensLeft;
+        }
+
+        tokensLeft.sub(tokens);
+
+        // Check if crowdsale is done.
+        if (tokensLeft == 0) {
+            nextState(State.Finished);
+        }
+
+        // Send tokens to buyer.
+        require(token.transfer(buyer, tokens));
+    }
+}
+
+contract Treasurer is Ownable {
+    enum State {
+        Initialized,
+        Started
+    }
+
+    address public team;
+    State public state;
+
+    GetToken token_;
+    Crowdsale crowdsale_;
+
+
+    uint256 public initialSupply   = 1000000 * 1000; // 1 million GET tokens.
+    uint256 public teamReserve     =  500000 * 1000; // 0.5 million GET tokens.
+    uint256 public crowdsaleAmount =  500000 * 1000; // 0.5 million GET tokens.
+
+    function Treasurer(address _team) public {
+        team = _team;
+        owner = msg.sender;
+        state = State.Initialized;
+    }
+
+    function start(IBosonAddressResolver _bosonResolver) onlyOwner public {
+        require(state == State.Initialized);
+        state = State.Started;
+
+        GetToken token = new GetToken(this, initialSupply);
+        Crowdsale crowdsale = new Crowdsale(token, _bosonResolver);
```

Comment:
This means we'll have to have whole code of Crowdsale included in this contract, wonder if we hit the gas limit